### ISSUE TITLE
Grain state lost when using multiple/generic grain interfaces

### ISSUE THREAD

**Issue by pentp** [dotnet/orleans#6308]
If a grain is deactivated, then its state can be lost if it's activated through a different interface depending on call order and if the GrainReference interface is generic or not.

Repro: https://github.com/pentp/GenericGrainRef/blob/master/Program.cs
Output:
```
OnActivateAsync GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
SetValue='42' GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
GetValue='42' GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
GetValue='42' GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
Deactivate GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
DeactivateActivationOnIdle: [Activation: S127.0.0.1:11111:319648827*grn/F09898EA/00000000@a08e3b87 #GrainType=ReproGrain Placement=RandomPlacement State=Deactivating] later when become idle.

Forwarding 1 requests destined for address S127.0.0.1:11111:319648827*grn/F09898EA/00000000@a08e3b87 to address  after Finished Destroy Activation.
Trying to forward after Finished Destroy Activation, ForwardCount = 0. OldAddress = S127.0.0.1:11111:319648827*grn/F09898EA/00000000@a08e3b87, ForwardingAddress = (null), Message Request S127.0.0.1:11111:319648827*cli/699e3387@1116aaab->S127.0.0.1:11111:319648827*grn/F09898EA/00000000@a08e3b87 #15: , Exception: (null).
OnActivateAsync GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
GetValue='42' GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
GetValue='42' GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
Deactivate GrainReference:*grn/F09898EA/00000000<[Dummy, GenericGrainRef, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]> Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
DeactivateActivationOnIdle: [Activation: S127.0.0.1:11111:319648827*grn/F09898EA/00000000@77542b91 #GrainType=ReproGrain Placement=RandomPlacement State=Deactivating] later when become idle.

Forwarding 1 requests destined for address S127.0.0.1:11111:319648827*grn/F09898EA/00000000@77542b91 to address  after Finished Destroy Activation.
Trying to forward after Finished Destroy Activation, ForwardCount = 0. OldAddress = S127.0.0.1:11111:319648827*grn/F09898EA/00000000@77542b91, ForwardingAddress = (null), Message Request S127.0.0.1:11111:319648827*cli/699e3387@1116aaab->S127.0.0.1:11111:319648827*grn/F09898EA/00000000@77542b91 #19: , Exception: (null).
OnActivateAsync GrainReference:*grn/F09898EA/00000000 Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
GetValue='' GrainReference:*grn/F09898EA/00000000 Identity=*grn/F09898EA/0000000000000000000000000000000003fffffff09898ea-0x544A6FCF
```
At this point the grain does not have its state anymore because `Grain.GrainReference` is different while `Identity` is the same as before.

This is really confusing and hard to debug because it depends on activation/call order. Grain state storage should not depend on the way it was activated (called) and Grain.GrainReference value should not change for the same grain.
#1123 might be relevant also.

**Comment by veikkoeeva** [comment]
@pentp It looks like https://github.com/dotnet/orleans/issues/1998 is the relevant issue.

**Comment by pentp** [comment]
Yes, that looks also relevant. But in this case the grains aren't even generic (they each have a concrete type that is also passed to the storage/reminder providers together with the GrainReference), they just implement a generic interface through which they can be activated. In this situation I would expected the internal GrainReference to not have any genericArguments.

**Comment by ReubenBond** [comment]
Fixed in 4.0. It is unlikely that we can backport the fix. It's related to how grains themselves are identified
